### PR TITLE
Feat/sqs index archive

### PR DESCRIPTION
**Pull request by sachitrajbhandari**
- Reads archive SQS messages and archive to dc.
- Reads index SQS messages, finds and download metadata yaml, index to dc.

### CODE REVIEW

File: odc_index/sqs_to_dc.py [hunk 1]
```diff
@@ -42,41 +49,58 @@ def queue_to_odc(
...
+                    logging.error("Archive skipped as failed to get ID")
                     ds_failed += 1
             else:
-                logging.error("Failed to get URI from metadata doc")
-                ds_failed += 1
+                metadata_uri = None
+                uri = None
+                for link in metadata.get("links"):
+                    rel = link.get("rel")
+                    if rel and rel == "derived_from":
```
Comment (alexgleith) [inline odc_index/sqs_to_dc.py:66]: Do you think that this is the right `rel` to use? The STAC Spec says it should be the parent `STAC` document, see: https://github.com/radiantearth/stac-spec/blob/dev/item-spec/item-spec.md#relation-types

I wonder if we should make this link explicitly something like `odc_yaml` so that it is unambiguous.

File: odc_index/sqs_to_dc.py [hunk 2]
```diff
@@ -42,41 +49,58 @@ def queue_to_odc(
...
+                        metadata_uri = link.get("href")
+                    elif rel and rel == "self":
+                        uri = link.get("href")
+
+                if metadata_uri:
+                    metadata = YAML().load(requests.get(metadata_uri).content)
```
Comment (alexgleith) [inline odc_index/sqs_to_dc.py:72]: If we're doing this, we shouldn't do a transform, I don't think.

The fact that we have an ODC YAML means that we should be able to assume that we can't do a transform.

File: odc_index/sqs_to_dc.py [hunk 3]
```diff
@@ -151,7 +179,7 @@ def cli(
 
     # Do the thing
     dc = Datacube()
-    added, failed = queue_to_odc(
+    added, updated, archived, failed = queue_to_odc(
```
Comment (alexgleith) [inline odc_index/sqs_to_dc.py:182]: I feel a bit weird about doing adding, updating and archiving in the same CLI, but it looks simple enough!

File: odc_index/sqs_to_dc.py (hunk 1 above)
Comment (sachitrajbhandari) [reply odc_index/sqs_to_dc.py]: Changed to `"rel": "odc_yaml"`

File: odc_index/sqs_to_dc.py (hunk 2 above)
Comment (sachitrajbhandari) [reply odc_index/sqs_to_dc.py:77]: Added `--odc-metadata-link` option to use ODC yaml link for indexing

**Comment by sachitrajbhandari** [comment]
Successfully tested loading of indexed dataset using `dc.load`

File: odc_index/sqs_to_dc.py [hunk 4]
```diff
@@ -42,41 +54,61 @@ def queue_to_odc(
...
+                uri = None
+                for link in metadata.get("links"):
+                    rel = link.get("rel")
+                    if odc_metadata_link and rel and rel == "odc_yaml":
+                        metadata_uri = link.get("href")
+                    elif rel and rel == "self":
+                        uri = link.get("href")
+
+                if metadata_uri:
+                    metadata = YAML().load(requests.get(metadata_uri).content)
```
Comment (Kirill888) [inline odc_index/sqs_to_dc.py:77]: 1. `requests.get(..)` can fail, this will stop entire process.
2. `YAML().load(..)` probably ends up using unsafe python version of YAML parsing that will be super slow and unsafe, how much work is this things supposed to be doing, a handful of documents or a million? It also can fail..

we should be using this for parsing yamls instead:
https://github.com/opendatacube/datacube-core/blob/f446c47f644cd5ab10b975a6675c66652481c02e/datacube/utils/documents.py#L62

**Comment by Kirill888** [comment]
I would recommend to split this giant monolithic function into several and add a lot more try/except to deal with possible failures.

You have several well defined stages in the data pipeline
1. Get SQS message
2. Fetch data from http/s3
3. Construct Dataset
4. Index/update Dataset

I would write those as separate Iterator -> Iterator transforms (using `yield`) and compose them that way. One single large while loop is to hard to follow and keep robust.

File: odc_index/sqs_to_dc.py [hunk 5]
```diff
@@ -42,41 +54,61 @@ def queue_to_odc(
...
+                    uri = metadata_uri
+                else:
+                    logging.error("ODC EO3 metadata link not found")
+
+                if transform:
+                    metadata = transform(metadata)
+
+                doc2ds = Doc2Dataset(dc.index, **kwargs)
```
Comment (Kirill888) [inline odc_index/sqs_to_dc.py:85]: Should not be creating this thing for every dataset processed, it actually queries database for products on construction. This object can and should be re-used across datasets.

File: odc_index/sqs_to_dc.py [hunk 6]
```diff
@@ -42,41 +54,61 @@ def queue_to_odc(
             message = messages[0]
             body = json.loads(message.body)
             metadata = json.loads(body["Message"])
+            if archive:
```
[the remark this answers is not in the document]
Comment (sachitrajbhandari) [reply odc_index/sqs_to_dc.py:57]: The UUID should be in the SNS message as the metadata file will be deleted from S3 in the archiving process.

File: odc_index/sqs_to_dc.py (hunk 5 above)
Comment (sachitrajbhandari) [reply odc_index/sqs_to_dc.py]: `Doc2Dataset` removed from loop and created only once.

File: odc_index/sqs_to_dc.py (hunk 4 above)
Comment (sachitrajbhandari) [reply odc_index/sqs_to_dc.py]: Exception handling added and using `documents.parse_yaml` from `datacube.util`

File: odc_index/sqs_to_dc.py [hunk 7]
```diff
@@ -161,10 +204,17 @@ def cli(
         transform=transform,
         limit=limit,
         update=update,
+        archive=archive,
         allow_unsafe=allow_unsafe,
+        odc_metadata_link=odc_metadata_link,
     )
 
-    print(f"Added {added} Datasets, Failed {failed} Datasets")
+    print(
```
[the remark this answers is not in the document]
Comment (sachitrajbhandari) [reply odc_index/sqs_to_dc.py]: Added action handler to show proper action result

File: odc_index/sqs_to_dc.py [hunk 8]
```diff
@@ -42,41 +54,61 @@ def queue_to_odc(
...
+                    if odc_metadata_link and rel and rel == "odc_yaml":
+                        metadata_uri = link.get("href")
+                    elif rel and rel == "self":
+                        uri = link.get("href")
+
+                if metadata_uri:
+                    metadata = YAML().load(requests.get(metadata_uri).content)
+                    uri = metadata_uri
+                else:
+                    logging.error("ODC EO3 metadata link not found")
```
[the remark this answers is not in the document]
Comment (sachitrajbhandari) [reply odc_index/sqs_to_dc.py]: Proper Exception handler added

File: odc_index/sqs_to_dc.py [hunk 9]
```diff
@@ -42,41 +49,58 @@ def queue_to_odc(
...
+                    rel = link.get("rel")
+                    if rel and rel == "odc_yaml":
+                        metadata_uri = link.get("href")
+                    elif rel and rel == "self":
+                        uri = link.get("href")
+
+                if metadata_uri:
+                    metadata = YAML().load(requests.get(metadata_uri).content)
+                    uri = metadata_uri
+                elif transform:
```
[the remark this answers is not in the document]
Comment (sachitrajbhandari) [reply odc_index/sqs_to_dc.py]: Fixed.

**Comment by sachitrajbhandari** [comment]
@alexgleith @Kirill888 The code has been refactored and address all the previous feedback. Can I have another round of review? Thanks.

File: odc_index/sqs_to_dc.py [hunk 10]
```diff
@@ -3,80 +3,182 @@
...
+            if archive:
+                # Archive metadata
+                do_archiving(metadata, dc)
             else:
-                logging.error("Failed to get URI from metadata doc")
-                ds_failed += 1
+                # Extract metadata and URI for indexing
+                metadata, uri = get_metadata_uri(metadata, transform, odc_metadata_link)
+                # Index metadata
+                do_indexing(metadata, uri, dc, doc2ds, update, allow_unsafe)
```
Comment (pindge) [inline odc_index/sqs_to_dc.py:70]: I need a condition here to handle indexing 

```Dockerfile
ADD https://raw.githubusercontent.com/opendatacube/datacube-dataset-config/master/scripts/index_from_s3_bucket.py /code
```

```python
from index_from_s3_bucket import add_dataset, get_s3_url
def process_message(index, message, prefix, sources_policy=DEFAULT_SOURCES_POLICY):
    # message body is a string, need to parse out json a few times
    inner = json.loads(message)
    s3_message = json.loads(inner["Message"])
    errors = dict()
    datasets = []
    skipped = 0
    if "Records" not in s3_message:
        errors["no_record"] = "Message did not contain S3 records"
        return datasets, errors

    for record in s3_message["Records"]:
        bucket_name = record["s3"]["bucket"]["name"]
        key = record["s3"]["object"]["key"]
        if prefix is None or len(prefix) == 0 or any([PurePath(key).match(p) for p in prefix]):
            try:
                errors[key] = None
                obj = s3.Object(bucket_name, key).get(
                    ResponseCacheControl='no-cache')
                data = load(obj['Body'].read())
                # NRT data may not have a creation_dt, attempt insert if missing
                if "creation_dt" not in data:
                    try:
                        data["creation_dt"] = data["extent"]["center_dt"]
                    except KeyError:
                        pass
                uri = get_s3_url(bucket_name, key)
                dataset, errors[key] = add_dataset(
                    data, uri, index, sources_policy)
                if errors[key] is None:
                    datasets.append(dataset)
            except Exception as e:
                errors[key] = e
        else:
            logger.debug("Skipped: %s as it does not match prefix filters", key)
            skipped = skipped + 1
    return datasets, skipped, errors
```

**Review by alexgleith — APPROVED** [review]
